File: Frontend/src/service/scenarioService.js
import BaseService from './baseService'

/**
 * Service class for handling scenario-related API operations
 * @extends BaseService
 */
class ScenarioService extends BaseService {
  /**
   * Creates an instance of ScenarioService with the scenarios API endpoint
   */
  constructor() {
    super('/scenarios')
  }

  /**
   * Retrieves all scenarios from the system
   * @async
   * @returns {Promise<Array>} Promise resolving to an array of scenario objects
   * @throws {Error} If the API request fails
   */
  async getAllScenarios() {
    try {
      return await this.get('')
    } catch (error) {
      console.error('[ScenarioService] Failed to fetch scenarios:', error)
      throw error
    }
  }
  /**
   * Retrieves a specific scenario by its ID
   * @async
   * @param {number} id - The unique identifier of the scenario to retrieve
   * @returns {Promise<Object>} Promise resolving to the scenario object
   * @throws {Error} If the API request fails
   */
  async getScenarioById(id) {
    try {
      return await this.get(`${id}`)
    } catch (error) {
      console.error('[ScenarioService] Failed to fetch scenario by ID:', error)
      throw error
    }
  }

  /**
   * Creates a new scenario in the system
   * @async
   * @param {Object} scenarioData - The data for the new scenario
   * @returns {Promise<Object>} Promise resolving to the API response with a success message
   * @throws {Error} If the scenario creation request fails
   */
  async createScenario(scenarioData) {
    try {
      return await this.post('', scenarioData)
    } catch (error) {
      console.error('[ScenarioService] Failed to create scenario:', error)
      throw error
    }
  }

  /**
   * Updates an existing scenario in the system
   * @async
   * @param {number} id - The unique identifier of the scenario to update
   * @param {Object} scenarioData - The updated scenario data
   * @returns {Promise<Object>} Promise resolving to the API response with a success message
   * @throws {Error} If the scenario update request fails
   */
  async updateScenario(id, scenarioData) {
    try {
      return await this.put(`${id}`, scenarioData)
    } catch (error) {
      console.error('[ScenarioService] Failed to update scenario:', error)
      throw error
    }
  }
}

export default new ScenarioService()

File: Backend/src/main/java/edu/ntnu/idatt2106/krisefikser/service/incident/ScenarioService.java
package edu.ntnu.idatt2106.krisefikser.service.incident;

import edu.ntnu.idatt2106.krisefikser.api.dto.incident.ScenarioRequestDto;
import edu.ntnu.idatt2106.krisefikser.api.dto.incident.ScenarioResponseDto;
import edu.ntnu.idatt2106.krisefikser.persistance.entity.incident.Scenario;
import edu.ntnu.idatt2106.krisefikser.persistance.repository.incident.ScenarioRepository;
import java.util.List;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

/**
 * Service for handling scenario related operations.
 */
@Service
public class ScenarioService {

  private static final Logger logger = LoggerFactory.getLogger(ScenarioService.class);
  private final ScenarioRepository scenarioRepository;

  /**
   * Constructor for ScenarioService.
   *
   * @param scenarioRepository The repository for scenario related operations.
   */
  public ScenarioService(ScenarioRepository scenarioRepository) {
    this.scenarioRepository = scenarioRepository;
  }

  /**
   * Creates a new scenario.
   *
   * @param dto The scenario request DTO
   */
  public void createScenario(ScenarioRequestDto dto) {
    if (dto.getName() == null || dto.getName().isBlank()) {
      logger.error("Scenario name is required");
      throw new IllegalArgumentException("Scenario name is required");
    }
    Scenario scenario = dto.toEntity();
    scenarioRepository.save(scenario);
    logger.info("Scenario created successfully: {}", scenario.getName());
  }

  /**
   * Updates an existing scenario by ID.
   *
   * @param id  The ID of the scenario to update
   * @param dto The updated scenario data
   */
  public void updateScenario(Long id, ScenarioRequestDto dto) {
    if (dto.getName() == null || dto.getName().isBlank()) {
      logger.error("Scenario name is required for update");
      throw new IllegalArgumentException("Scenario name is required");
    }

    Scenario scenario = scenarioRepository.findById(id)
        .orElseThrow(() -> {
          logger.warn("Scenario not found with ID: {}", id);
          return new IllegalArgumentException("Scenario not found with ID: " + id);
        });

    scenario.setName(dto.getName());
    scenario.setDescription(dto.getDescription());
    scenario.setToDo(dto.getToDo());
    scenario.setPackingList(dto.getPackingList());
    scenario.setIconName(dto.getIconName());

    scenarioRepository.save(scenario);
    logger.info("Scenario with ID {} updated successfully", id);
  }

  /**
   * Retrieves all scenarios.
   *
   * @return A list of all scenarios
   */
  public List<ScenarioResponseDto> getAllScenarios() {
    return scenarioRepository.findAll().stream()
        .map(ScenarioResponseDto::fromEntity)
        .toList();
  }

  /**
   * Retrieves a scenario by ID.
   *
   * @param id The ID of the scenario to retrieve
   * @return The scenario response DTO
   */
  public ScenarioResponseDto getScenarioById(Long id) {
    Scenario scenario = scenarioRepository.findById(id)
        .orElseThrow(() -> {
          logger.warn("Scenario not found with ID: {}", id);
          return new IllegalArgumentException("Scenario not found with ID: " + id);
        });
    return ScenarioResponseDto.fromEntity(scenario);
  }
}
